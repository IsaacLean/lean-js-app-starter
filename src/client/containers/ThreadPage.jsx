import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import Loading from '../components/Loading'
import Nav from '../components/Nav'
import Threads from '../components/Threads'
import { endFetch, fetchThreads, setErr } from '../actions'
import { getThread } from '../reducers'
import { setDocTitle } from '../util'

class ThreadPage extends Component {
  componentDidMount() {
    if(this.props.thread) {
      setDocTitle(this.props.thread.subject || `Thread ${this.props.thread._id}`)
    }

    this.props.fetchThreads(this.props.id)
      .then(thread => {
        setDocTitle(thread.subject || `Thread ${thread._id}`)
      })
      .catch(err => {
        if(!this.props.thread) {
          this.props.setErr('read', err)
        } else {
          this.props.endFetch()
        }
      })
  }

  render() {
    let content
    if(this.props.thread) {
      content = <Threads threads={[this.props.thread]} />
    } else if(!this.props.thread && this.props.isFetching) {
      content = <Loading />
    }

    return <>
      <Nav mode="top" />
      {content}
      <Nav mode="bottom" />
    </>
  }
}

ThreadPage.propTypes = {
  endFetch: PropTypes.func.isRequired,
  fetchThreads: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  isFetching: PropTypes.bool.isRequired,
  setErr: PropTypes.func.isRequired,
  thread: PropTypes.object
}

const mapStateToProps = (state, ownProps) => ({
  isFetching: state.isFetching,
  thread: getThread(state, ownProps.id)
})

const mapDispatchToProps = dispatch => bindActionCreators({
  endFetch,
  fetchThreads,
  setErr
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ThreadPage)
