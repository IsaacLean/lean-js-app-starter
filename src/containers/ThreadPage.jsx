import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import Nav from '../components/Nav'
import Threads from '../components/Threads'
import { fetchThreads } from '../actions'
import { getThread } from '../reducers'
import { setDocTitle } from '../util'

class ThreadPage extends Component {
  componentDidMount() {
    if(!this.props.thread) {
      this.props.fetchThreads(this.props.id)
        .then(thread => {
          setDocTitle(thread.subject || `Thread ${thread._id}`)
        })
    } else {
      setDocTitle(this.props.thread.subject || `Thread ${this.props.thread._id}`)
    }
  }

  render() {
    let thread = []
    if(this.props.thread) {
      thread.push(this.props.thread)
    }

    return <>
      <Nav mode="top" />
      <Threads threads={thread} />
      <Nav mode="bottom" />
    </>
  }
}

ThreadPage.propTypes = {
  fetchThreads: PropTypes.func,
  id: PropTypes.string,
  thread: PropTypes.object
}

const mapStateToProps = (state, ownProps) => ({
  thread: getThread(state, ownProps.id)
})

const mapDispatchToProps = dispatch => bindActionCreators({ fetchThreads }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ThreadPage)
