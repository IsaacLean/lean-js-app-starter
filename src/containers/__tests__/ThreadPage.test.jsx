import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'

import { ThreadPage } from '../ThreadPage'

Enzyme.configure({ adapter: new Adapter() })

describe('ThreadPage', () => {
  const data = {
    _id: '5bee2f90c43e83433b5e5d24',
    subject: 'New Thread',
    comment: 'I am the new thread comment.',
    createdAt: '2018-11-16T02:46:40.301Z',
    replies: [
      {
        _id: '5bee2f90c43e83433b5e5d25',
        createdAt: '2018-11-16T02:46:40.311Z',
        comment: 'All your base are belong to us.',
        type: 'Reply'
      }
    ],
    type: 'Thread'
  }

  it('renders properly when loading', () => {
    const component = shallow(<ThreadPage id={data._id} isFetching={true} />)
    expect(component).toMatchSnapshot()
  })

  it('renders properly when loaded', () => {
    const component = shallow(<ThreadPage id={data._id} isFetching={false} thread={data} />)
    expect(component).toMatchSnapshot()
  })
})
