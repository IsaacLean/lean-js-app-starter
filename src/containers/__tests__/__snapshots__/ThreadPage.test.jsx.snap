// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadPage renders properly 1`] = `
<Fragment>
  <Nav
    mode="top"
  />
  <Loading />
  <Nav
    mode="bottom"
  />
</Fragment>
`;

exports[`ThreadPage renders properly when loaded 1`] = `
<Fragment>
  <Nav
    mode="top"
  />
  <Threads
    threads={
      Array [
        Object {
          "_id": "5bee2f90c43e83433b5e5d24",
          "comment": "I am the new thread comment.",
          "createdAt": "2018-11-16T02:46:40.301Z",
          "replies": Array [
            Object {
              "_id": "5bee2f90c43e83433b5e5d25",
              "comment": "All your base are belong to us.",
              "createdAt": "2018-11-16T02:46:40.311Z",
              "type": "Reply",
            },
          ],
          "subject": "New Thread",
          "type": "Thread",
        },
      ]
    }
  />
  <Nav
    mode="bottom"
  />
</Fragment>
`;

exports[`ThreadPage renders properly when loading 1`] = `
<Fragment>
  <Nav
    mode="top"
  />
  <Loading />
  <Nav
    mode="bottom"
  />
</Fragment>
`;
